src/torrent.cpp

```diff
@@ -5291,12 +5291,20 @@ namespace libtorrent
 		TORRENT_ASSERT(std::count(m_peers_to_disconnect.begin()
 			, m_peers_to_disconnect.end(), p) == 0);
 
-		std::weak_ptr<torrent> weak_t = shared_from_this();
-		m_peers_to_disconnect.push_back(p);
-		m_deferred_disconnect.post(m_ses.get_io_service(), [=]() {
-			std::shared_ptr<torrent> t = weak_t.lock();
-			if (t) t->on_remove_peers();
-		});
+		// we only schedule the peer for actual removal if in fact
+		// it is attached to this torrent, since this method can
+		// be called from the attach_peer path and fail to do so
+		// because of too many connections
+		if (p->associated_torrent().lock().get() == this)
```

Comment:
I don't really understand this comment. My understanding is that the peer ``p`` is part of the ``m_connections`` vector at this point, no? Is that not a prerequisite for this function?
would it be a more accurate test, to not add the peer to the deferred list if it hasn't been added to that list? iirc, m_connections is sorted, and fairly cheap to do a lookup in